lib/Conversion/ImportVerilog/Expressions.cpp

```diff
@@ -1171,9 +1196,91 @@ struct RvalueExprVisitor : public ExprVisitor {
+
+    // Qualified call: t.m(...), extract from thisClass.
+    if (const slang::ast::Expression *recvExpr = expr.thisClass()) {
+      thisRef = context.convertRvalueExpression(*recvExpr);
+      if (!thisRef)
+        return {};
+
+      handleTy = dyn_cast<moore::ClassHandleType>(thisRef.getType());
+      if (!handleTy) {
+        mlir::emitError(loc)
+            << "receiver of method '" << expr.getSubroutineName()
+            << "' must be class<...>, got " << thisRef.getType();
+        return {};
+      }
+    } else {
+      // Unqualified call inside a method body: try using implicit %this.
+      thisRef = context.getImplicitThisRef();
+      if (!thisRef) {
+        mlir::emitError(loc) << "method '" << expr.getSubroutineName()
+                             << "' called without an object";
+        return {};
+      }
+      handleTy = dyn_cast<moore::ClassHandleType>(thisRef.getType());
+      if (!handleTy) {
+        mlir::emitError(loc)
+            << "implicit 'this' must be class<...>, got " << thisRef.getType();
+        return {};
+      }
+    }
+    return {thisRef, handleTy};
+  }
+
+  /// Build either a class.vcall or class.call for a method call
+  mlir::Value buildMethodCall(const slang::ast::SubroutineSymbol *subroutine,
+                              FunctionLowering *lowering,
+                              moore::ClassHandleType actualHandleTy,
+                              mlir::Value actualThisRef,
+                              SmallVector<Value> &arguments,
+                              //! This does NOT include the implicit handle!
+                              SmallVector<Type> &resultTypes) {
+
+    // Get the expected receiver type from the lowered method
+    auto funcTy = lowering->op.getFunctionType();
+    auto expected0 = funcTy.getInput(0);
+    auto expectedHdlTy = cast<moore::ClassHandleType>(expected0);
+
+    // Upcast the handle as necessary.
+    mlir::Value implicitThisRef = context.materializeConversion(
+        expectedHdlTy, actualThisRef, false, actualThisRef.getLoc());
+
+    // Build an argument list where the this reference is the first argument.
+    SmallVector<Value> explicitArguments;
+    explicitArguments.push_back(implicitThisRef);
+    explicitArguments.append(arguments.begin(), arguments.end());
+
+    // Method call: choose direct vs virtual.
+    const bool isVirtual =
+        (subroutine->flags & slang::ast::MethodFlags::Virtual) != 0;
+
+    if (!isVirtual) {
+      // Direct (non-virtual) call -> moore.class.call
+      auto calleeSym = mlir::SymbolRefAttr::get(context.getContext(),
+                                                lowering->op.getSymName());
+      auto callOp = mlir::func::CallOp::create(builder, loc, resultTypes,
+                                               calleeSym, explicitArguments);
+      return resultTypes.size() > 0 ? callOp.getResult(0) : Value{};
+    }
+
+    mlir::emitError(loc) << "Virtual methods are not yet supported!";
```

Comment:
Nit: the frontend tries to stick with lowercase error messages:
```suggestion
    mlir::emitError(loc) << "virtual method calls not supported";
```